File: packages/go/body_contract_test.go
package crumbtrail

import (
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

// Conformance against test-fixtures/backend-body/cases.json.
//
// The Ruby and ASP.NET Core packages run the same file. Reading it from the repository root
// rather than copying it in is the whole point: three hand written copies of this policy is
// how Ruby, Go and .NET ended up redacting the same body three different ways.
type bodyCase struct {
	Name  string          `json:"name"`
	Why   string          `json:"why"`
	Input string          `json:"input"`
	State string          `json:"state"`
	Body  json.RawMessage `json:"body"`
}

type bodyCorpus struct {
	Policy string `json:"policy"`
	Limits struct {
		Bytes         int   `json:"bytes"`
		Nesting       int   `json:"nesting"`
		Keys          int   `json:"keys"`
		Items         int   `json:"items"`
		IntegerDigits int   `json:"integerDigits"`
		SafeInteger   int64 `json:"safeInteger"`
	} `json:"limits"`
	Cases []bodyCase `json:"cases"`
}

func repositoryRoot(t *testing.T) string {
	t.Helper()
	dir, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	for {
		if _, err := os.Stat(filepath.Join(dir, "test-fixtures")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			t.Fatal("repository root not found")
		}
		dir = parent
	}
}

func loadBodyCorpus(t *testing.T) bodyCorpus {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join(repositoryRoot(t), "test-fixtures", "backend-body", "cases.json"))
	if err != nil {
		t.Fatal(err)
	}
	var corpus bodyCorpus
	if err := json.Unmarshal(raw, &corpus); err != nil {
		t.Fatal(err)
	}
	if len(corpus.Cases) == 0 {
		// Without this every assertion below would pass vacuously.
		t.Fatal("backend body corpus is empty")
	}
	return corpus
}

func decode(t *testing.T, raw []byte) any {
	t.Helper()
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.UseNumber()
	var value any
	if err := decoder.Decode(&value); err != nil {
		t.Fatalf("cannot decode %s: %v", raw, err)
	}
	return value
}

func TestBodyCorpusLimits(t *testing.T) {
	corpus := loadBodyCorpus(t)
	if corpus.Policy != policy {
		t.Fatalf("policy %q, corpus says %q", policy, corpus.Policy)
	}
	for _, check := range []struct {
		name           string
		actual, corpus int64
	}{
		{"bytes", bodyLimit, int64(corpus.Limits.Bytes)},
		{"nesting", maxNesting, int64(corpus.Limits.Nesting)},
		{"keys", maxKeys, int64(corpus.Limits.Keys)},
		{"items", maxItems, int64(corpus.Limits.Items)},
		{"integerDigits", maxIntegerDigits, int64(corpus.Limits.IntegerDigits)},
		{"safeInteger", safeInteger, corpus.Limits.SafeInteger},
	} {
		if check.actual != check.corpus {
			t.Errorf("%s limit is %d, corpus says %d", check.name, check.actual, check.corpus)
		}
	}
}

func TestBodyCorpus(t *testing.T) {
	for _, example := range loadBodyCorpus(t).Cases {
		t.Run(example.Name, func(t *testing.T) {
			got := captureBody([]byte(example.Input), false)
			if got.State != example.State {
				t.Fatalf("state %q, want %q: %s", got.State, example.State, example.Why)
			}
			if len(example.Body) == 0 || string(example.Body) == "null" {
				if got.Body != nil {
					t.Fatalf("body %v should have been withheld entirely", got.Body)
				}
				return
			}
			body, ok := got.Body.(string)
			if !ok {
				t.Fatalf("body %v is not serialized JSON", got.Body)
			}
			if want, actual := decode(t, example.Body), decode(t, []byte(body)); !reflect.DeepEqual(want, actual) {
				t.Fatalf("body %s, want %s: %s", body, example.Body, example.Why)
			}
		})
	}
}

File: packages/ruby/test/body_contract_test.rb
require 'minitest/autorun'
require 'json'
require 'crumbtrail'

# Conformance against `test-fixtures/backend-body/cases.json`.
#
# The Go and ASP.NET Core packages run the same file. Reading it from the repository root
# rather than copying it in is the whole point: three hand written copies of this policy is
# how Ruby, Go and .NET ended up redacting the same body three different ways.
class BodyContractTest < Minitest::Test
  CORPUS = begin
    directory = File.expand_path(__dir__)
    directory = File.dirname(directory) until File.exist?(File.join(directory, 'test-fixtures/backend-body/cases.json')) || directory == '/'
    path = File.join(directory, 'test-fixtures/backend-body/cases.json')
    raise "backend body corpus not found from #{__dir__}" unless File.exist?(path)
    JSON.parse(File.read(path))
  end

  def test_corpus_is_reachable
    # Without this every other assertion here could pass vacuously.
    refute_empty CORPUS['cases']
    assert_equal Crumbtrail::Body::POLICY, CORPUS['policy']
  end

  def test_limits_match_the_corpus
    limits = CORPUS['limits']
    assert_equal limits['bytes'], Crumbtrail::Body::LIMIT
    assert_equal limits['nesting'], Crumbtrail::Body::MAX_NESTING
    assert_equal limits['keys'], Crumbtrail::Body::MAX_KEYS
    assert_equal limits['items'], Crumbtrail::Body::MAX_ITEMS
    assert_equal limits['integerDigits'], Crumbtrail::Body::MAX_INTEGER_DIGITS
    assert_equal limits['safeInteger'], Crumbtrail::Body::SAFE_INTEGER
  end

  CORPUS['cases'].each do |example|
    define_method("test_#{example['name']}") do
      body, state = Crumbtrail::Body.capture(example['input'].b)
      assert_equal example['state'], state, "#{example['name']}: #{example['why']}"
      if example['body'].nil?
        assert_nil body, "#{example['name']} should withhold the body entirely"
      else
        assert_equal example['body'], JSON.parse(body), "#{example['name']}: #{example['why']}"
      end
    end
  end
end
